### PR TITLE
fix out of index in deser

### PR DESCRIPTION
**Pull request by longbowlu**

### CODE REVIEW

**Review by velvia — APPROVED** [review]
👍

**Review by kchalkias — APPROVED** [review]
This fix is correct, we should enforce that we received at least as many bytes as the struct/enum name (which already includes `::`) + at least one byte (which would define the payload).

File: crates/sui-types/src/crypto.rs [hunk 1]
```diff
@@ -1368,6 +1368,9 @@ where
         // Remove name tag before deserialization using BCS
         let name = serde_name::trace_name::<Self>().expect("Self should be a struct or an enum");
         let name_byte_len = format!("{}::", name).bytes().len();
+        if name_byte_len >= bytes.len() {
+            anyhow::bail!("Failed to deserialize signable bytes {name}");
```
Comment (kchalkias) [inline crates/sui-types/src/crypto.rs:1372]: For logging purposes I'd also add that `shorter that expected payload was received` and we might attach the length received (`bytes.len()`).

File: crates/sui-types/src/crypto.rs [hunk 2]
```diff
@@ -1368,6 +1368,9 @@ where
         // Remove name tag before deserialization using BCS
         let name = serde_name::trace_name::<Self>().expect("Self should be a struct or an enum");
         let name_byte_len = format!("{}::", name).bytes().len();
```
Comment (kchalkias) [inline crates/sui-types/src/crypto.rs:1370]: Also a short optimization for the future is that to compute `name_len_byte` we could just do `name.len() + 2`, no need to rewrite it

**Pull request merged**